lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp

```diff
@@ -0,0 +1,204 @@
+    // shorter paths.
+    llvm::DenseMap<SILBasicBlock*, BBState> visitedBBs;
+    SmallVector<SILBasicBlock*, 16> worklist;
+
+    auto *entryBB = &(*fun.begin());
+    visitedBBs.try_emplace(entryBB);
+    worklist.push_back(entryBB);
+
+    // Track that last return instruction seen by the analysis, if any, before
+    // encountering a yield.
+    // This is needed for emitting diagnostics when there is no yield
+    // instruction along any path reaching a return instruction.
+    llvm::Optional<SILInstruction *> returnInst = None;
+    ASTContext &astCtx = fun.getModule().getASTContext();
+
+    while (!worklist.empty()) {
+      SILBasicBlock *bb = worklist.pop_back_val();
+      BBState state = visitedBBs[bb];
+
+      auto *term = bb->getTerminator();
+      if (isa<ReturnInst>(term)) {
+        // A conflict state implies that there is a path to return before
+        // seeing a yield.
+        if (state.yieldState == YieldState::Conflict) {
+          diagnose(astCtx, term->getLoc().getSourceLoc(),
+                   diag::possible_return_before_yield);
+          // Add a note to the instruction where the conflict first appeared.
+          diagnose(astCtx,
+                   state.conflictInst.getValue()->getLoc().getSourceLoc(),
+                   diag::conflicting_join);
+          return;
+        }
+        // If the state is BeforeYield, it is an error. But, defer emitting
+        // diagnostics until we see a Conflict state or have visited all nodes
+        // in order to gather more information.
+        if (state.yieldState != YieldState::AfterYield) {
+          returnInst = term;
+        }
+        continue;
+      }
+
+      // Check whether there are multiple yields.
+      if (isa<YieldInst>(term)) {
+        if (state.yieldState != YieldState::BeforeYield) {
+          diagnose(astCtx, term->getLoc().getSourceLoc(),
+                   diag::multiple_yields);
+          // Add a note that points to the previous yield.
+          diagnose(astCtx, state.yieldInst.getValue()->getLoc().getSourceLoc(),
+                   diag::previous_yield);
+          return;
+        }
+      }
+
+      for (auto &succ : term->getSuccessors()) {
+        SILBasicBlock *succBB = succ.getBB();
+
+        // Optimistically try to set our current state as the state
+        // of the successor.
+        auto insertResult = visitedBBs.try_emplace(succBB, state);
+
+        // If the insertion was successful, it means we are seeing the successor
+        // for the first time. Propogate the state and add the successor to the
+        // worklist.
+        if (insertResult.second) {
+          worklist.insert(worklist.begin(), succBB);
+
+          // When the successor follows a 'yield', update the successor state.
+          if (isa<YieldInst>(term)) {
+            insertResult.first->second.yieldState = YieldState::AfterYield;
+            insertResult.first->second.yieldInst = term;
```

Comment:
You should consider encapsulating `BBState`: add functions to do the various state transitions (these can assert that (1) the current state is appropriate for the transition and (2) they've been provided the expected information (e.g. a yield instruction)) and functions to get the state-specific information (these can assert that the state is as expected).  This sort of approach encourages you to write better, stronger assertions and can catch problems a lot faster if the code needs to evolve.